### ISSUE TITLE
[BUG]: Two method are not existing for zarinpal driver, and only work on idpay.

### ISSUE THREAD

**Issue by jalallinux** [evryn/laravel-toman#21]
Hello @AmirrezaNasiri  ✌🏻

**Describe the bug**
Two method are not existing for zarinpal driver,  and only work on idpay.

**To Reproduce**
1. Set default gateway to zarinpal in config/toman.php
2. Create Toman request:
```php
 Toman::amount(10000)
            ->orderId('123')
            ->mobile('a valid mobile number')
            ->email('a valid email')
            ->callback('a valid site route')
            ->description('description of this payment')
            ->name('JalalLinuX')
            ->request();
```
3. See exception:
```
Method Evryn\\LaravelToman\\PendingRequest::orderId does not exist.
```
4. Now comment ->orderId('123')
```php
 Toman::amount(10000)
            // ->orderId('123')
            ->mobile('a valid mobile number')
            ->email('a valid email')
            ->callback('a valid site route')
            ->description('description of this payment')
            ->name('JalalLinuX')
            ->request();
```
5. See exception:
```
Method Evryn\\LaravelToman\\PendingRequest::name does not exist.
```

**Expected behavior**
This package a multi pay driver package and it should have the same usage. but `orderId`, `name` method only work for idpay driver.

**Screenshots**
<img width="991" alt="Screen Shot 1400-07-22 at 17 21 16" src="https://user-images.githubusercontent.com/37062636/137331036-215e7f84-c200-4d96-bf9f-14eb51a74e47.png">
<img width="995" alt="Screen Shot 1400-07-22 at 17 26 13" src="https://user-images.githubusercontent.com/37062636/137331964-51f43fea-b411-4b07-8ff4-72503ddaa39b.png">

**Exact Versions:**
 - Laravel: 8.54
 - PHP: 7.4
 - Package: 2.0.1

**Comment by AmirrezaNasiri** [comment]
Hi @jalallinux :)

Thanks for the issue submission. This is actually a discussable topic; The way we should generalize the package API and I've struggled with the issue when I was working on it.

I chose to follow driver-specific requirements for requests and callbacks and yet generalize some common parts of them; So developers can understand which driver expects what kind of inputs and considerations explicitly. This can for instance prevent depending on the non-existence `$request->orderId()` in the callback requests when the gateway has switched from IDPay to Zarinpal - because there is no definition of `orderId` in Zarinpal.
A developer can also use their own strategy design pattern to make gateways interchangeable easily. So my +1 goes for this architecture.

On the other hand, it's about the easiness of usage. A similar package for the study is [Laravel Socialite](https://socialiteproviders.com/about/), which generalizes callbacks (same in our case) and puts its related logic up to the developer itself. Though its providers are not completely interchangeable either (because of lack of the common attributes), and we also have the same issue for requesting side, I guess it's a +1 for this architecture too.

To summarize, I'm not sure about either of them. Developers need their own logic for interchangeability for both of them. I'm gonna keep this issue open - This is an interesting topic.
